### PR TITLE
url: fix .title vs url callback plugins

### LINKED ISSUE #2054: url: `exclude` setting also applies to `.title` command
### Description
Before Sopel 7.1, the only sane way to disable auto-title would have been setting the config option `url.exclude` to `.*` (matching any possible URL). The only problem is, doing that _also_ makes `.title` not work.

At least as far back as Willie 4.0.0, when the responsible code was restructured (in d1acb126a51a6353bd559ab1d9b36c1e4b9e5c83), and probably even longer, doing e.g. `.title http://thisurlis.excluded` would yield silence from the bot.

Recently, the silence has changed to an error message, since it's impossible for the code added in #1380 (then re-added in #2029) to tell the difference between a failed fetch, an excluded link, or a link handled by some other plugin.

### Reproduction steps
1. Set `url.exclude` to `.*`
2. Run `.title http://any.link/of/your/choosing`
3. Get an error: "Sorry, fetching that title failed. Make sure the site is working."

### Expected behavior
I expect Sopel to try and fetch the link's title anyway, because it's been explicitly requested.

### Environment
- Sopel `.version`: 7.1.0-dev
- Sopel installed via: `pip install -e`
- Python version: 3.8.5
- Operating system: Ubuntu 18.04
- Relevant plugins: url

### Notes
Since this behavior has been around for _so long_, I elected to hold any change until 8.0, even though any fix is probably trivial enough to include in 7.1 still. We'd be changing the meaning of a setting.

It seems easiest to fix both this and #2055 at the same time.

### LINKED ISSUE #2055: url: `.title` command shows an error for links handled by another plugin
### Description
Links handled by another plugin (for example links to GitHub issues, if the `github` plugin is installed) are skipped intentionally in the `.title` command, to avoid giving duplicate output. However, the error message code from #2029 can't tell the difference between a link that was intentionally skipped and a link for which a title couldn't be fetched.

### Reproduction steps
1. Command Sopel, `.title http://some.link/that/another/plugin/handles`
   - With core plugins only, a Reddit post is a good way to trigger this.
2. `url` responds with an error: "Sorry, fetching that title failed. Make sure the site is working."
   AND the plugin that knows how to handle the link will also reply

### Expected behavior
The `url` plugin should be silent in this case, unless multiple URLs are passed to `.title` and at least one of them isn't handled by another plugin.

### Environment
- Sopel `.version`: 7.1.0-dev
- Sopel installed via: `pip install -e`
- Python version: 3.8.5
- Operating system: Ubuntu 18.04
- Relevant plugins: url

### Notes
Likely easiest to fix this in conjunction with #2054. I'm therefore marking this for 8.0 even though it would be nice to fix it in 7.1. (The target milestone is, as always, subject to change after discussion.)

### PR DESCRIPTION
**Pull request by half-duplex**
- .title for e.g. reddit would say "Sorry, couldn't title that" followed by the reddit.py callback output
- last_seen_url (therefore `.title` sans args) was missing callback'd/excluded URLs
- tighten some security dumbs (except the safety part in #2279)
  - no point in checking ip without DNS lookups if an attacker has $5 and a namecheap account
- add type hints
- some other tidying.

Related for some remaining issues: #2281
Fixes #2054
Fixes #2055

### Checklist
- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

### CODE REVIEW

File: sopel/modules/url.py [hunk 1]
```diff
@@ -154,7 +149,7 @@ def shutdown(bot):
 @plugin.example('.urlpexclude example\\.com/\\w+', user_help=True)
 @plugin.example('.urlexclude example.com/path', user_help=True)
 @plugin.output_prefix('[url] ')
-def url_ban(bot, trigger):
+def url_ban(bot: Sopel, trigger: Trigger):
```
Comment (Exirel) [inline sopel/modules/url.py:152]: This is not an instance of `Sopel` but of `SopelWrapper`, true for the other commands & rules.

Comment (half-duplex) [reply sopel/modules/url.py:152]: Ah, thanks. IIRC in the safety PR you said to use either - same there? Or what's different?

Comment (Exirel) [reply sopel/modules/url.py:152]: Plugin callables need a SopelWrapper. Job, Setup, Teardown: Sopel. Everything else: it depends.

**Comment by half-duplex** [comment]
RE_DCC - Seems to strip "dcc send". Should this ever matter? DCC is CTCP only, and title responses should never be sent as CTCP, right?

**Comment by Exirel** [comment]
Given that you are fixing an issues, that's pretty impressive work. I like that every time you work on something, you make it more clear for me and give me better ideas than I had for future refactoring.

So all in all, good job.

File: sopel/modules/url.py [hunk 2]
```diff
@@ -311,55 +303,68 @@ def title_auto(bot, trigger):
...
+
+    If a callback was dispatched, only the url and dispatched=True will be set.
+
+    For titles explicitly requested by the user, exclusion_char and excludes
+    are skipped.
+
+    :param bot: Sopel instance
+    :param trigger: The trigger object for this event
+    :param urls: The URLs detected in the triggering message
+    :param requested: Whether the title was explicitly requested (vs automatic)
```
Comment (Exirel) [inline sopel/modules/url.py:336]: The format used in other part of the code is usually in that order:

* one line synopsis
* parameters
* long description

So this bloc of 4 lines should be above the long description, and not at the end.

**Review by Exirel — CHANGES_REQUESTED** [review]
Mostly style nitpicks. All the complains I have about `process_urls` come from the complexity of the feature as outgrow the old interface that was put in place (probably by me at some point).

So this is now in my "refactoring" bucket, and clearly not in the near future because, as I already said: this PR is a damn good work.

**Comment by Exirel** [comment]
Is there a reason for this PR to still be a draft? It looks pretty good to me. Did I forget something?

**Comment by half-duplex** [comment]
I was going to remove the DCC thing if we can't figure out why it's there, but I guess that's stretching the scope a bit.

**Review by Exirel — APPROVED** [review]

**Comment by dgw** [comment]
I wouldn't mind removing the seemingly useless DCC regex, but I'll merge this regardless of whether that's included by the time I get to its place in the test/review queue.

**Comment by dgw** [comment]
> by the time I get to its place in the test/review queue.

I did. It passed my gauntlet. 🎉

**Pull request merged**